File: backend/app/schemas/discovery.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from app.schemas.base import CamelModel


class DiscoverySourceRead(CamelModel):
    key: str
    label: str
    provider: str
    enabled: bool
    configured: bool
    notes: str | None = None


class DiscoverySourceListResponse(CamelModel):
    items: list[DiscoverySourceRead]


class DiscoveryCandidateRead(CamelModel):
    id: UUID
    wallet_address: str
    source: str
    source_rank: int | None
    source_label: str | None
    source_cohort: str | None
    account_value: Decimal | None
    source_pnl: Decimal | None
    source_roi: Decimal | None
    source_copy_score: Decimal | None
    account_role: str
    parent_address: str | None
    subaccount_name: str | None
    status: str
    fail_reason: str | None
    backfill_status: str
    backfill_error: str | None
    last_backfilled_at: datetime | None
    backfill_fetched_count: int
    backfill_inserted_count: int
    backfill_duplicate_count: int
    fill_count: int
    closed_trade_count: int
    open_trade_count: int
    ignored_fill_count: int
    net_pnl_usd: Decimal | None
    profit_factor: Decimal | None
    win_rate: Decimal | None
    max_drawdown_pct: Decimal | None
    average_trade_notional_usd: Decimal | None
    last_trade_time_ms: int | None
    first_seen_at: datetime
    last_seen_at: datetime
    created_at: datetime
    updated_at: datetime


class DiscoveryCandidateListResponse(CamelModel):
    items: list[DiscoveryCandidateRead]
    total: int
    limit: int
    offset: int


class DiscoveryImportRunRead(CamelModel):
    id: UUID
    source: str
    status: str
    requested_limit: int
    fetched_count: int
    candidate_count: int
    inserted_count: int
    updated_count: int
    skipped_count: int
    error: str | None
    started_at: datetime
    finished_at: datetime | None


class DiscoveryImportRunListResponse(CamelModel):
    items: list[DiscoveryImportRunRead]
    total: int
    limit: int
    offset: int


class DiscoveryPrefilterResponse(CamelModel):
    evaluated: int
    accepted: int
    rejected: int
    unchanged: int
    reject_reasons: dict[str, int]
    candidates: list[DiscoveryCandidateRead]


class DiscoveryBackfillItem(CamelModel):
    wallet_address: str
    source: str
    status: str
    fail_reason: str | None
    pool_action: str | None = None
    fetched: int
    inserted: int
    duplicate: int
    fill_count: int
    closed_trade_count: int
    open_trade_count: int
    ignored_fill_count: int
    net_pnl_usd: Decimal | None
    profit_factor: Decimal | None
    win_rate: Decimal | None
    max_drawdown_pct: Decimal | None
    error: str | None = None


class DiscoveryBackfillResponse(CamelModel):
    scanned: int
    backfilled: int
    accepted: int
    rejected: int
    promoted: int
    pool_inserted: int
    pool_duplicate: int
    failed: int
    skipped: int
    reject_reasons: dict[str, int]
    items: list[DiscoveryBackfillItem]


class DiscoveryPromoteItem(CamelModel):
    wallet_address: str
    source: str
    action: str
    label: str | None
    already_in_pool: bool
    reason: str | None = None


class DiscoveryPromoteResponse(CamelModel):
    scanned: int
    promoted: int
    inserted: int
    duplicate: int
    skipped: int
    items: list[DiscoveryPromoteItem]


class DiscoveryImportResponse(CamelModel):
    requested_sources: list[str]
    limit: int
    runs: list[DiscoveryImportRunRead]
    candidates: list[DiscoveryCandidateRead]
    fetched: int
    candidate_count: int
    inserted: int
    updated: int
    skipped: int
    skip_reasons: dict[str, int]
    failed_sources: int
    prefilter: DiscoveryPrefilterResponse | None = None
    backfill: DiscoveryBackfillResponse | None = None

File: frontend/src/types/discovery.ts
export type DiscoverySource = {
  key: string;
  label: string;
  provider: string;
  enabled: boolean;
  configured: boolean;
  notes: string | null;
};

export type DiscoverySourceListResponse = {
  items: DiscoverySource[];
};

export type DiscoveryCandidate = {
  id: string;
  walletAddress: string;
  source: string;
  sourceRank: number | null;
  sourceLabel: string | null;
  sourceCohort: string | null;
  accountValue: string | null;
  sourcePnl: string | null;
  sourceRoi: string | null;
  sourceCopyScore: string | null;
  accountRole: string;
  parentAddress: string | null;
  subaccountName: string | null;
  status: string;
  failReason: string | null;
  backfillStatus: string;
  backfillError: string | null;
  lastBackfilledAt: string | null;
  backfillFetchedCount: number;
  backfillInsertedCount: number;
  backfillDuplicateCount: number;
  fillCount: number;
  closedTradeCount: number;
  openTradeCount: number;
  ignoredFillCount: number;
  netPnlUsd: string | null;
  profitFactor: string | null;
  winRate: string | null;
  maxDrawdownPct: string | null;
  averageTradeNotionalUsd: string | null;
  lastTradeTimeMs: number | null;
  firstSeenAt: string;
  lastSeenAt: string;
  createdAt: string;
  updatedAt: string;
};

export type DiscoveryCandidateListResponse = {
  items: DiscoveryCandidate[];
  total: number;
  limit: number;
  offset: number;
};

export type DiscoveryImportRun = {
  id: string;
  source: string;
  status: string;
  requestedLimit: number;
  fetchedCount: number;
  candidateCount: number;
  insertedCount: number;
  updatedCount: number;
  skippedCount: number;
  error: string | null;
  startedAt: string;
  finishedAt: string | null;
};

export type DiscoveryImportRunListResponse = {
  items: DiscoveryImportRun[];
  total: number;
  limit: number;
  offset: number;
};

export type DiscoveryPrefilterResponse = {
  evaluated: number;
  accepted: number;
  rejected: number;
  unchanged: number;
  rejectReasons: Record<string, number>;
  candidates: DiscoveryCandidate[];
};

export type DiscoveryBackfillItem = {
  walletAddress: string;
  source: string;
  status: string;
  failReason: string | null;
  poolAction: string | null;
  fetched: number;
  inserted: number;
  duplicate: number;
  fillCount: number;
  closedTradeCount: number;
  openTradeCount: number;
  ignoredFillCount: number;
  netPnlUsd: string | null;
  profitFactor: string | null;
  winRate: string | null;
  maxDrawdownPct: string | null;
  error: string | null;
};

export type DiscoveryBackfillResponse = {
  scanned: number;
  backfilled: number;
  accepted: number;
  rejected: number;
  promoted: number;
  poolInserted: number;
  poolDuplicate: number;
  failed: number;
  skipped: number;
  rejectReasons: Record<string, number>;
  items: DiscoveryBackfillItem[];
};

export type DiscoveryPromoteItem = {
  walletAddress: string;
  source: string;
  action: string;
  label: string | null;
  alreadyInPool: boolean;
  reason: string | null;
};

export type DiscoveryPromoteResponse = {
  scanned: number;
  promoted: number;
  inserted: number;
  duplicate: number;
  skipped: number;
  items: DiscoveryPromoteItem[];
};

export type DiscoveryImportResponse = {
  requestedSources: string[];
  limit: number;
  runs: DiscoveryImportRun[];
  candidates: DiscoveryCandidate[];
  fetched: number;
  candidateCount: number;
  inserted: number;
  updated: number;
  skipped: number;
  skipReasons: Record<string, number>;
  failedSources: number;
  prefilter: DiscoveryPrefilterResponse | null;
  backfill: DiscoveryBackfillResponse | null;
};

export type DiscoveryActionResponse =
  | DiscoveryImportResponse
  | DiscoveryPrefilterResponse
  | DiscoveryBackfillResponse
  | DiscoveryPromoteResponse;
